node/src/components/chain_synchronizer/operations.rs

```diff
@@ -351,125 +357,184 @@ impl CanUseSyncingNodes for Deploy {}
-async fn get_filtered_fully_connected_peers<REv>(ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
+/// Possible errors caused by fetch operation that uses the retry mechanism.
+#[derive(Error, Debug)]
+pub(crate) enum FetchWithRetryError<T>
 where
-    REv: From<NetworkInfoRequest>,
+    T: Item,
 {
-    let mut peer_list = ctx.effect_builder.get_fully_connected_peers().await;
-    ctx.filter_bad_peers(&mut peer_list);
-    peer_list
+    #[error(
+        "Fetch attempts exhausted for item with id {id:?}. Total attempts: {total_attempts}, \
+        attempts while bootstrapped: {attempts_after_bootstrapped}"
+    )]
+    AttemptsExhausted {
+        id: T::Id,
+        total_attempts: usize,
+        attempts_after_bootstrapped: usize,
+    },
+
+    #[error(transparent)]
+    FetcherError(#[from] FetcherError<T>),
 }
 
-/// Fetches an item. Keeps retrying to fetch until it is successful. Not suited to fetching a block
-/// header or block by height, which require verification with finality signatures.
-async fn fetch_retry_forever<REv, T>(ctx: &ChainSyncContext<'_, REv>, id: T::Id) -> FetchResult<T>
+/// Fetches an item.
+///
+/// Not suited to fetching a block header or block by height, which require verification with
+/// finality signatures.
+async fn fetch_with_retries<REv, T>(
+    ctx: &ChainSyncContext<'_, REv>,
+    id: T::Id,
+) -> Result<FetchedData<T>, FetchWithRetryError<T>>
 where
     T: Item + CanUseSyncingNodes + 'static,
     REv: From<FetcherRequest<T>> + From<NetworkInfoRequest>,
 {
-    let mut attempts = 0_usize;
+    let mut total_attempts = 0;
+    let mut attempts_after_bootstrapped = 0;
     loop {
-        let new_peer_list = if T::can_use_syncing_nodes() {
-            get_filtered_fully_connected_peers(ctx).await
-        } else {
-            get_filtered_fully_connected_non_syncing_peers(ctx).await
-        };
-
-        if new_peer_list.is_empty() && attempts % 100 == 0 {
+        let has_connected_to_network = has_connected_to_network();
+        let new_peer_list = get_peers(T::can_use_syncing_nodes(), ctx).await;
+        if new_peer_list.is_empty() && total_attempts % 100 == 0 {
             warn!(
-                attempts,
+                total_attempts,
+                attempts_after_bootstrapped,
+                has_connected_to_network,
                 item_type = ?T::TAG,
                 ?id,
                 can_use_syncing_nodes = %T::can_use_syncing_nodes(),
                 "failed to attempt to fetch item due to no fully-connected peers"
             );
         }
 
-        for peer in new_peer_list {
-            trace!(
-                "attempting to fetch {:?} with id {:?} from {:?}",
-                T::TAG,
+        if let Some(value) = fetch_from_peers(new_peer_list, id, ctx).await {
+            return value;
+        }
+
+        total_attempts += 1;
+        if has_connected_to_network {
+            attempts_after_bootstrapped += 1;
+        }
```

Comment:
Why do we track # of fetch requests before and after connecting? Why is that important?

Reply:
It's a convenience for potentially easier debugging.